File: behavioral-design-patterns/chain-of-responsibility-design-pattern/Python/swiggy-order.py
class OrderHandler:
    def __init__(self, next_handler) -> None:
        self.next_handler: OrderHandler = next_handler
    
    def process_order(self, order: str):
        pass

class OrderValidationHandler(OrderHandler):
    def __init__(self, next_handler) -> None:
        super().__init__(next_handler)
        
    def process_order(self, order: str):
        print(f"Validating order: {order}")
        
        if self.next_handler is not None:
            self.next_handler.process_order(order=order)

class PaymentProcessingHandler(OrderHandler):
    def __init__(self, next_handler) -> None:
        super().__init__(next_handler)
        
    def process_order(self, order: str):
        print(f"Processing payment for order: {order}")
        
        if self.next_handler is not None:
            self.next_handler.process_order(order=order)

class OrderPreparationHandler(OrderHandler):
    def __init__(self, next_handler) -> None:
        super().__init__(next_handler)
        
    def process_order(self, order: str):
        print(f"Preparing order: {order}")
        
        if self.next_handler is not None:
            self.next_handler.process_order(order=order)

class DeliveryAssignmentHandler(OrderHandler):
    def __init__(self, next_handler) -> None:
        super().__init__(next_handler)
        
    def process_order(self, order: str):
        print(f"Assigning delivery for order: {order}")
        
        if self.next_handler is not None:
            self.next_handler.process_order(order=order)

class OrderTrackingHandler(OrderHandler):
    def __init__(self, next_handler) -> None:
        super().__init__(next_handler)
        
    def process_order(self, order: str):
        print(f"Tracking order: {order}")
        
        if self.next_handler is not None:
            self.next_handler.process_order(order=order)
            
class SwiggyOrder:
    @staticmethod
    def main():
        order_processing_chain: OrderHandler = OrderValidationHandler(
            PaymentProcessingHandler(
                OrderPreparationHandler(
                    DeliveryAssignmentHandler(
                        OrderTrackingHandler(None)                   
                    )
                )
            )
        )
        
        order: str = "Pizza"
        order_processing_chain.process_order(order=order)
        
SwiggyOrder.main()

File: behavioral-design-patterns/chain-of-responsibility-design-pattern/Typescript/swiggy-order.ts
class OrderHandler {
    public nextHandler: OrderHandler | null;
    constructor (nextHandler: OrderHandler | null) {
        this.nextHandler = nextHandler;
    }
    public processOrder(order: string) {}
}

class OrderValidationHandler extends OrderHandler {
    constructor(nextHandler: OrderHandler | null) {
        super(nextHandler);
    }

    public processOrder(order: string) {
        console.log(`Validating order: ${order}`);

        if (this.nextHandler != null) {
            this.nextHandler.processOrder(order=order)
        }
    }
}
class PaymentProcessingHandler extends OrderHandler {
    constructor(nextHandler: OrderHandler | null) {
        super(nextHandler);
    }

    public processOrder(order: string) {
        console.log(`Processing payment for order: ${order}`);

        if (this.nextHandler != null) {
            this.nextHandler.processOrder(order=order)
        }
    }
}
class OrderPreparationHandler extends OrderHandler {
    constructor(nextHandler: OrderHandler | null) {
        super(nextHandler);
    }

    public processOrder(order: string) {
        console.log(`Preparing order: ${order}`);

        if (this.nextHandler != null) {
            this.nextHandler.processOrder(order=order)
        }
    }
}
class DeliveryAssignmentHandler extends OrderHandler {
    constructor(nextHandler: OrderHandler | null) {
        super(nextHandler);
    }

    public processOrder(order: string) {
        console.log(`Assigning delivery for order: ${order}`);

        if (this.nextHandler != null) {
            this.nextHandler.processOrder(order=order)
        }
    }
}
class OrderTrackingHandler extends OrderHandler {
    constructor(nextHandler: OrderHandler | null) {
        super(nextHandler);
    }

    public processOrder(order: string) {
        console.log(`Tracking order: ${order}`);

        if (this.nextHandler != null) {
            this.nextHandler.processOrder(order=order)
        }
    }
}

class SwiggyOrder {
    public static main() {
        const orderProcessingChain: OrderHandler | null = new OrderValidationHandler(
            new PaymentProcessingHandler(
                new OrderPreparationHandler(
                    new DeliveryAssignmentHandler(
                        new OrderTrackingHandler(null)
                    )
                )
            )
        )

        const order: string = "Pizza";
        orderProcessingChain.processOrder(order);
    }
}

SwiggyOrder.main()